### PR TITLE
Add cluster join command line options and configuration options

### PR DESCRIPTION
**Pull request by ChrisAubuchon**
* Adds `-rejoin`, `-join`, `-retry-join`, `-retry-max` and `-retry-interval` command line options
* Adds `start_join`, `retry_join`, `retry_max`, `retry_interval` and `rejoin_after_leave` server configuration options

### CODE REVIEW

**Comment by cbednarski** [comment]
Thanks for the PR. Can you also update the website documentation under `website`?

**Comment by ChrisAubuchon** [comment]
I updated `config.html.md`. Is there anywhere else that needs to be changed?

File: command/agent/config.go [hunk 1]
```diff
@@ -358,14 +388,6 @@ func (a *Config) Merge(b *Config) *Config {
 		result.AdvertiseAddrs = result.AdvertiseAddrs.Merge(b.AdvertiseAddrs)
 	}
 
-	// Apply the Atlas configuration
```
Comment (cbednarski) [inline command/agent/config.go]: Why was this removed?

File: command/agent/config_test.go [hunk 2]
```diff
@@ -114,6 +109,11 @@ func TestConfig_Merge(t *testing.T) {
 			NumSchedulers:     2,
 			EnabledSchedulers: []string{structs.JobTypeBatch},
 			NodeGCThreshold:   "12h",
+			RejoinAfterLeave:  true,
```
Comment (cbednarski) [inline command/agent/config_test.go]: Is there a test case that runs without any of these options set?

**Comment by cbednarski** [comment]
> I updated config.html.md. Is there anywhere else that needs to be changed?

Looks great! That's probably the only spot in the docs that needs an update.

File: command/agent/config.go (hunk 1 above)
Comment (ChrisAubuchon) [inline command/agent/config.go]: Bad git'ing on my part.

File: command/agent/config_test.go (hunk 2 above)
Comment (ChrisAubuchon) [inline command/agent/config_test.go]: The only tests that use these options are TestConfig_Merge, TestConfig_LoadConfigString, and TestRetryJoin. Is there a specific test case that you are thinking of?

File: command/agent/config.go (hunk 1 above)
Comment (cbednarski) [inline command/agent/config.go]: :+1:

File: command/agent/config_test.go (hunk 2 above)
Comment (cbednarski) [inline command/agent/config_test.go]: > Is there a specific test case that you are thinking of?

No, I just like to make sure we test the default case where none of the new options is specified so we don't get a regression at some point. If there is not an existing test case that covers this we should add a new one. I asked because it's hard to see this from the diff; I will look more closely later.

**Comment by c4milo** [comment]
I'm looking forward to having this PR merge to make cluster bootstrapping more automatic.

File: command/agent/command.go [hunk 3]
```diff
@@ -100,6 +105,15 @@ func (c *Command) readConfig() *Config {
 		return nil
 	}
 
+	if cmdConfig.Server.RetryInterval != "" {
+		dur, err := time.ParseDuration(cmdConfig.Server.RetryInterval)
+		if err != nil {
+			c.Ui.Error(fmt.Sprintf("Error: %s", err))
```
Comment (diptanu) [inline command/agent/command.go]: Can we add some extra information in the error message to indicate which field Nomad was not about to parse? I think that would help the user to quickly debug what's wrong in the configuration file.

**Comment by diptanu** [comment]
Merging this PR into a branch to do more testing and understand how the UX works.

File: command/agent/command.go [hunk 4]
```diff
@@ -396,12 +416,16 @@ func (c *Command) Run(args []string) int {
 	// Enable log streaming
 	logGate.Flush()
 
+	// Start retry join process
+	errCh := make(chan struct{})
```
Comment (dadgar) [inline command/agent/command.go]: Can we add the retryJoinErr channel to the Command struct so we don't have to pass the channel into the functions

Comment (ChrisAubuchon) [inline command/agent/command.go]: Added to the Command struct

File: command/agent/command.go (hunk 3 above)
Comment (ChrisAubuchon) [inline command/agent/command.go]: I updated the error message to `Error parsing retry interval`

**Comment by dadgar** [comment]
Thanks @ChrisAubuchon
